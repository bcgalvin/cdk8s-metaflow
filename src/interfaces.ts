import { HttpIngressPathType, ImagePullPolicy, ServiceType } from './enums';
import { Affinity, EnvFromSource, IngressTls, SecurityContext, Toleration } from './imports/k8s';

export interface ImageOptions {
  readonly repository?: string;
  readonly pullPolicy?: ImagePullPolicy;
  readonly tag?: string;
}

export interface ServiceAccountOptions {
  readonly create?: boolean | undefined;
  readonly annotations?: Record<string, string>;
  readonly name?: string;
}

export interface DbMigrationOptions {
  readonly runOnStart?: boolean;
  readonly onlyIfDbEmpty?: boolean;
}

export interface ServiceOptions {
  readonly type?: ServiceType;
  readonly port?: number;
  readonly annotations?: Record<string, string>;
}

export interface HttpIngressPath {
  readonly path?: string;
  readonly pathType?: HttpIngressPathType;
}

export interface HostOptions {
  readonly host?: string;
  readonly paths?: HttpIngressPath[];
}

export interface IngressOptions {
  readonly enabled?: boolean;
  readonly className?: string;
  readonly annotations?: Record<string, string>;
  readonly hosts?: HostOptions[];
  readonly tls?: IngressTls;
}

export interface AutoscalingOptions {
  readonly enabled?: boolean;
  readonly minReplicas?: number;
  readonly maxReplicas?: number;
  readonly targetCPUUtilizationPercentage?: number;
}

export interface IApiResource {
  readonly apiGroup?: string;
  readonly resourceType?: string;
  readonly resourceName?: string;
}

export interface MetadataDatabaseEnvOptions {
  readonly port?: number;
  readonly password?: string;
  readonly host?: string;
  readonly user?: string;
}

export interface MetaflowServiceOptions {
  readonly replicaCount?: number;
  readonly image?: ImageOptions;
  readonly imagePullSecrets?: string[];
  readonly nameOverride?: string;
  readonly fullnameOverride?: string;
  readonly serviceAccount?: ServiceAccountOptions;
  readonly podAnnotations?: Record<string, string>;
  readonly podSecurityContext?: Record<string, string>;
  readonly dbMigrations?: DbMigrationOptions;
  readonly securityContext?: SecurityContext;
  readonly service?: ServiceOptions;
  readonly ingress?: IngressOptions;
  readonly resources?: IApiResource[];
  readonly autoscaling?: AutoscalingOptions;
  readonly nodeSelector?: Record<string, string>;
  readonly tolerations?: Toleration[];
  readonly affinity?: Affinity;
  readonly envFrom?: EnvFromSource[];
  readonly metadatadb?: MetadataDatabaseEnvOptions;
}

export interface MetaflowUIOptions {
  readonly replicaCount?: number;
  readonly image?: ImageOptions;
  readonly uiImage?: ImageOptions;
  readonly imagePullSecrets?: string[];
  readonly nameOverride?: string;
  readonly fullnameOverride?: string;
  readonly env: Record<string, string>;
  readonly serviceAccount?: ServiceAccountOptions;
  readonly podAnnotations?: Record<string, string>;
  readonly podSecurityContext?: Record<string, string>;
  readonly securityContext?: SecurityContext;
  readonly service?: ServiceOptions;
  readonly serviceStatic?: ServiceOptions;
  readonly ingress?: IngressOptions;
  readonly resources?: IApiResource[];
  readonly autoscaling?: AutoscalingOptions;
  readonly nodeSelector?: Record<string, string>;
  readonly tolerations?: Toleration[];
  readonly affinity?: Affinity;
  readonly envFrom?: EnvFromSource[];
  readonly metadatadb?: MetadataDatabaseEnvOptions;
  readonly metaflowDatastoreSysrootS3?: string;
}

export interface DatabaseAuthOptions {
  readonly database?: string;
  readonly password?: string;
  readonly username?: string;
  readonly postgresPassword?: string;
  readonly enablePostgresUser?: boolean;
  readonly replicationPassword?: string;
  readonly replicationUsername?: string;
}

export interface DatabaseMetricsOptions {
  readonly enabled?: boolean;
}

export interface DatabaseResourceRequestOptions {
  readonly memory?: string;
  readonly cpu?: string;
}

export interface DatabaseResourcesOptions {
  readonly requests?: DatabaseResourceRequestOptions;
}

export interface DatabaseReplicationOptions {
  readonly enabled?: boolean;
  readonly readReplicas?: number;
}

export interface DatabaseVolumePermissionsOptions {
  readonly enabled?: boolean;
}

export interface MetadataDatabaseOptions {
  readonly auth?: DatabaseAuthOptions;
  readonly metrics?: DatabaseMetricsOptions;
  readonly resources?: DatabaseResourcesOptions;
  readonly replication?: DatabaseReplicationOptions;
  readonly architecture?: string;
  readonly volumePermissions?: DatabaseVolumePermissionsOptions;
}
