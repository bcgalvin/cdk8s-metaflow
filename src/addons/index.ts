export * from './addon';
export * from './postgres';
export * from './argo-workflows';
export * from './autoscaler';
