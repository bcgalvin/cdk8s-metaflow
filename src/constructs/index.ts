export * from './metaflow-service';
export * from './interfaces';
