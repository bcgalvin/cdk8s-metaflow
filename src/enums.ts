export enum ImagePullPolicy {
  ALWAYS = 'Always',
  IF_NOT_PRESENT = 'IfNotPresent',
  NEVER = 'Never',
}

export enum HttpIngressPathType {
  PREFIX = 'Prefix',
  EXACT = 'Exact',
  IMPLEMENTATION_SPECIFIC = 'ImplementationSpecific',
}

export enum ServiceType {
  CLUSTER_IP = 'ClusterIP',
  NODE_PORT = 'NodePort',
  LOAD_BALANCER = 'LoadBalancer',
  EXTERNAL_NAME = 'ExternalName',
}
