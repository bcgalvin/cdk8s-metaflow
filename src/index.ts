export * from './constructs';
export * from './addons';
