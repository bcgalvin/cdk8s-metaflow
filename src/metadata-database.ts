import { Helm } from 'cdk8s';
import { Construct } from 'constructs';
import { MetadataDatabaseOptions } from './interfaces';

export interface MetadataDatabaseProps {
  readonly chartVersion: string;
  readonly chartValues?: MetadataDatabaseOptions;
}

export class MetadataDatabase extends Construct {
  constructor(scope: Construct, id: string, props: MetadataDatabaseProps) {
    super(scope, id);

    const chartValues = props.chartValues ?? {
      enabled: true,
      auth: {
        database: 'metaflow',
        password: 'metaflow',
        username: 'metaflow',
      },
    };

    new Helm(this, 'postgres', {
      chart: 'bitnami/postgresql',
      releaseName: 'release-name-postgresql',
      helmFlags: ['--version', props.chartVersion],
      values: chartValues,
    });
  }
}
