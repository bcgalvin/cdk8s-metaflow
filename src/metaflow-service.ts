import * as kplus from 'cdk8s-plus-22';
import { Protocol } from 'cdk8s-plus-22';
import { Construct } from 'constructs';
import { KubeServiceAccount } from './imports/k8s';

export interface MetaflowServiceProps {
  readonly serviceAccountName?: string;
  readonly metadataServicePort?: number;
  readonly upgradesServicePort?: number;
}

export class MetaflowService extends Construct {
  constructor(scope: Construct, id: string, props?: MetaflowServiceProps) {
    super(scope, id);

    const metadataServicePort = props?.metadataServicePort || 8080;
    const upgradesServicePort = props?.upgradesServicePort || 8082;
    const serviceAccountName = props?.serviceAccountName || 'release-name-metaflow-service';

    new KubeServiceAccount(this, 'metaflow-sa', {
      metadata: {
        name: serviceAccountName,
        labels: {
          'helm.sh/chart': 'metaflow-service-0.2.0',
          'app.kubernetes.io/name': 'metaflow-service',
          'app.kubernetes.io/instance': 'release-name',
          'app.kubernetes.io/version': '2.2.4',
          'app.kubernetes.io/managed-by': 'Helm',
        },
      },
    });

    new kplus.Service(this, 'metaflow-service', {
      metadata: {
        name: serviceAccountName,
        labels: {
          'helm.sh/chart': 'metaflow-service-0.2.0',
          'app.kubernetes.io/name': 'metaflow-service',
          'app.kubernetes.io/instance': 'release-name',
          'app.kubernetes.io/version': '2.2.4',
          'app.kubernetes.io/managed-by': 'Helm',
        },
      },
      type: kplus.ServiceType.CLUSTER_IP,
      ports: [
        {
          name: 'metadata',
          port: metadataServicePort,
          protocol: Protocol.TCP,
          targetPort: metadataServicePort,
        },
        {
          name: 'upgrades',
          port: upgradesServicePort,
          protocol: Protocol.TCP,
          targetPort: upgradesServicePort,
        },
      ],
    });
  }
}
