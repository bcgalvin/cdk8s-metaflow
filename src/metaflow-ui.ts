import { Helm } from 'cdk8s';
import { Construct } from 'constructs';

export interface MetaflowUIProps {
  readonly chartVersion: string;
}

export class MetaflowUI extends Construct {
  constructor(scope: Construct, id: string, props: MetaflowUIProps) {
    super(scope, id);

    new Helm(this, 'metaflow-ui-helm', {
      chart: 'metaflow/metaflow-ui',
      releaseName: 'release-name-metaflow-ui',
      helmFlags: ['--version', props.chartVersion],
    });
  }
}
