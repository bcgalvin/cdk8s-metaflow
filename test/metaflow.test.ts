import { Chart, Testing } from 'cdk8s';
import { MetaflowService } from '../src';

const metadata = {
  'helm.sh/chart': 'metaflow-service-0.2.0',
  'app.kubernetes.io/name': 'metaflow-service',
  'app.kubernetes.io/instance': 'release-name',
  'app.kubernetes.io/version': '2.2.4',
  'app.kubernetes.io/managed-by': 'Helm',
};

describe('metaflow service', () => {
  test('defaults', () => {
    // GIVEN
    const app = Testing.app();
    const chart = new Chart(app, 'default');
    // WHEN
    new MetaflowService(chart, 'default-service');
    // THEN
    expect(Testing.synth(chart)).toMatchSnapshot();
  });

  test('defaults can be overridden', () => {
    // GIVEN
    const app = Testing.app();
    const chart = new Chart(app, 'override');
    // WHEN
    new MetaflowService(chart, 'override-service', {
      serviceAccountName: 'override-sa',
    });
    // THEN
    const resources = Testing.synth(chart);
    for (const sa of resources.filter((r) => r.kind === 'ServiceAccount')) {
      expect(sa.metadata.name).toEqual('override-sa');
    }
    for (const svc of resources.filter((r) => r.kind === 'Service')) {
      expect(svc.metadata.name).toEqual('override-sa');
    }
  });

  test('metadata', () => {
    // GIVEN
    const app = Testing.app();
    const chart = new Chart(app, 'metadata');
    // WHEN
    new MetaflowService(chart, 'metadata-service');
    // THEN
    const resources = Testing.synth(chart);
    for (const sa of resources.filter((r) => r.kind === 'ServiceAccount')) {
      expect(sa.metadata.labels).toEqual(metadata);
    }
    for (const svc of resources.filter((r) => r.kind === 'Service')) {
      expect(svc.metadata.labels).toEqual(metadata);
    }
  });
});
