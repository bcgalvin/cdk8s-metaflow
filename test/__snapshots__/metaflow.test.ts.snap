// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metaflow service defaults 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "ServiceAccount",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "release-name",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "metaflow-service",
        "app.kubernetes.io/version": "2.2.4",
        "helm.sh/chart": "metaflow-service-0.2.0",
      },
      "name": "release-name-metaflow-service",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "release-name",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "metaflow-service",
        "app.kubernetes.io/version": "2.2.4",
        "helm.sh/chart": "metaflow-service-0.2.0",
      },
      "name": "release-name-metaflow-service",
    },
    "spec": Object {
      "externalIPs": Array [],
      "ports": Array [
        Object {
          "name": "metadata",
          "port": 8080,
          "protocol": "TCP",
          "targetPort": 8080,
        },
        Object {
          "name": "upgrades",
          "port": 8082,
          "protocol": "TCP",
          "targetPort": 8082,
        },
      ],
      "selector": Object {},
      "type": "ClusterIP",
    },
  },
]
`;
